nexus/src/external_api/http_entrypoints.rs

```diff
@@ -921,13 +921,23 @@ impl NexusExternalApi for NexusExternalApiImpl {
         new_project: TypedBody<params::ProjectCreate>,
     ) -> Result<HttpResponseCreated<Project>, HttpError> {
         let apictx = rqctx.context();
-        let nexus = &apictx.context.nexus;
         let handler = async {
             let opctx =
                 crate::context::op_context_for_external_api(&rqctx).await?;
-            let project =
-                nexus.project_create(&opctx, &new_project.into_inner()).await?;
-            Ok(HttpResponseCreated(project.into()))
+            let nexus = &apictx.context.nexus;
+            let audit = nexus.audit_log_entry_init(&opctx, &rqctx).await?;
+
+            let result = async {
+                let project = nexus
+                    .project_create(&opctx, &new_project.into_inner())
+                    .await?;
+                Ok(HttpResponseCreated(project.into()))
+            }
+            .await;
+
+            let _ =
+                nexus.audit_log_entry_complete(&opctx, &audit, &result).await;
+            result
```

Comment:
It seems to me that, in absence of `dropshot` itself having a notion of middleware, a nicer way to make it easy to add audit logging to every API endpoint without having to repeat the `audit_log_entry_init`/`audit_log_entry_complete` calls everywhere is to have something similar to `ExternalLatencies::instrument_dropshot_handler`. This would just take an `async` block and a `rqctx`, and call the init function, run the rest of the handler, and then call the `complete` function. It could be composed with the rest of the stuff similarly to `instrument_dropshot_handler`.

For most API requests, we could even just have one such wrapper function thingy that does both the audit log stuff and the external-latencies instrumentation.

Reply:
Yes, I think that's right. One interesting bit is that if we want to record something like the ID of the resource created, we will have to figure out how to extract that generically from the inner block. Maybe it's not that hard, maybe we just always return a tuple of the response and then whatever bit of data want to log, and we just require that it be serializable, either as a string or as a json blob.

Reply:
as we discussed on Matrix, the current behavior is that if an API request is interrupted (i.e. by the client giving up or that Nexus crashing) but a saga starts successfully and eventually creates a resource, the completion of the audit log entry for that request will be "unknown"...which is not the worst thing, but might be nice to improve later. we can punt on that for the time being.